crates/miden-tx/src/tests/kernel_tests/test_fpi.rs

```diff
@@ -654,50 +659,66 @@ fn test_fpi_execute_foreign_procedure() {
 // NESTED FPI TESTS
 // ================================================================================================
 
-/// Test the correctness of the nested foreign procedures call.
+/// Test the correctness of the cyclic foreign procedure calls.
 ///
 /// It checks that the account data pointers are correctly added and removed from the account data
 /// stack.
 ///
 /// The call chain in this test looks like so:
-/// `Native -> First FA -> Second FA`
+/// `Native -> First FA -> Second FA -> First FA`
 #[test]
-fn test_simple_nested_fpi() {
+fn test_nested_fpi_cyclic_invocation() {
     // ------ SECOND FOREIGN ACCOUNT ---------------------------------------------------------------
-    let storage_slots = vec![AccountStorage::mock_item_0().slot];
     let second_foreign_account_code_source = "
-        use.miden::account
+        use.miden::tx
+        
+        use.std::sys
 
-        export.get_item_foreign
-            # make this foreign procedure unique to make sure that we invoke the procedure of the 
-            # foreign account, not the native one
-            push.1 drop
-            exec.account::get_item
+        export.second_account_foreign_proc
+            # get the storage item at index 0
+            # pad the stack for the `execute_foreign_procedure` execution
+            padw padw padw push.0.0
+            # => [pad(14)]
 
-            # return the first element of the resulting word
-            drop drop drop
+            # get the hash of the `get_item_foreign` account procedure from the advice stack
+            adv_push.4
```

Comment:
Do you think it might be easier to understand if in the first account we retrieved the `procref.get_item_foreign` as well as the account ID and then pass those 6 felts on the stack as arguments to `second_account_foreign_proc`? That way we wouldn't have to access the advice stack and it's easier to understand which procedure we're calling than with `first_foreign_account.code().procedures()[1].mast_root()`

Reply:
It will work, but because we need to provide the IDs and roots of all accounts and procedures we want to execute, the maximum length of such circle will be two (`first -> second -> third -> first`), just because every invocation needs 6 felts, and we can use at most 15 stack elements as inputs. And also potentially we won't be able to provide the necessary parameters to the final procedure since too many slots will be used by account and procedure data. So as far as I see using the advice stack is the only option which allows us to use cycles freely.

Reply:
Right, good point. Using the advice stack is less limiting than passing these as parameters, so that may be a better approach for reference of pioneers, for example.